runtime/vm/bindnatv.cpp

```diff
@@ -1095,29 +1090,34 @@ lookupJNINative(J9VMThread *currentThread, J9NativeLibrary *nativeLibrary, J9Met
 	PORT_ACCESS_FROM_JAVAVM(vm);
 
 	Trc_VM_lookupJNINative_Entry(currentThread, nativeLibrary, nativeMethod, symbolName, signature);
-#if JAVA_SPEC_VERSION >= 15
+#if JAVA_SPEC_VERSION >= 17
 	if (NULL == nativeLibrary) {
-		J9MemoryManagerFunctions const * const mmFuncs = vm->memoryManagerFunctions;
-		J9NameAndSignature nas = {0};
-		nas.name = (J9UTF8 *)&j9_findnative_name;
-		nas.signature = (J9UTF8 *)&j9_findnative_sig;
 		internalAcquireVMAccess(currentThread);
-		j9object_t entryName = mmFuncs->j9gc_createJavaLangString(currentThread, (U_8*)symbolName, strlen(symbolName), 0);
-		j9object_t classLoaderObject = J9_CLASS_FROM_METHOD(nativeMethod)->classLoader->classLoaderObject;
-		UDATA args[] = { (UDATA) classLoaderObject, (UDATA) entryName };
-		runStaticMethod(currentThread, (U_8 *)"java/lang/ClassLoader", &nas, 2, (UDATA *)args);
-		internalReleaseVMAccess(currentThread);
-		functionAddress = (UDATA *) currentThread->returnValue;
-		if ((NULL != currentThread->currentException) || (NULL == functionAddress) ) {
+		j9object_t entryName = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, (U_8*)symbolName, strlen(symbolName), 0);
+		if (NULL != entryName) {
+			j9object_t classLoaderObject = J9_CLASS_FROM_METHOD(nativeMethod)->classLoader->classLoaderObject;
+			J9Method *findNativeMethod = J9VMJAVALANGCLASSLOADER_FINDNATIVE_METHOD(vm);
+			UDATA args[] = {(UDATA)classLoaderObject, (UDATA)entryName};
+			internalRunStaticMethod(currentThread, findNativeMethod, TRUE, (sizeof(args) / sizeof(UDATA)), args);
+			internalReleaseVMAccess(currentThread);
+			functionAddress = (UDATA*)(*(U_64*)&(currentThread->returnValue));
+		}
+		if (NULL != currentThread->currentException) {
```

Comment:
There are VMHelper functions for checking/clearing which should be used. Also, rather than reacquiring VM access, just do these checks immediately upon return (functionAddress is guaranteed to be NULL in the exception case).

Reply:
To cover both exception cases, just move this up above the release VM access.

Reply:
You could unconditionally clear the exception and just use 0 == functionAddress to detect failure.

Reply:
See above about clearing unconditionally.

Reply:
Yes, `restoreCallInFrame` ensures it.

Reply:
Always clear the pending exception, check `functionAddress` value only for success.